resolver/resolver.go

```diff
@@ -41,25 +41,42 @@ func (e ErrNoLink) Error() string {
 	return fmt.Sprintf("no link named %q under %s", e.Name, e.Node.String())
 }
 
-// Resolver provides path resolution to IPFS
+// Resolver provides path resolution to IPFS.
+type Resolver interface {
+	// ResolveToLastNode walks the given path and returns the cid of the
+	// last block referenced by the path, and the path segments to
+	// traverse from the final block boundary to the final node within the
+	// block.
+	ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.Cid, []string, error)
+	// ResolvePath fetches the node for given path. It returns the last
+	// item returned by ResolvePathComponents and the last link traversed
+	// which can be used to recover the block.
+	ResolvePath(ctx context.Context, fpath path.Path) (ipld.Node, ipld.Link, error)
+	// ResolvePathComponents fetches the nodes for each segment of the given path.
+	// It uses the first path component as a hash (key) of the first node, then
+	// resolves all other components walking the links via a selector traversal
+	ResolvePathComponents(ctx context.Context, fpath path.Path) ([]ipld.Node, error)
+}
+
+// BasicResolver implements the Resolver interface.
 // It references a FetcherFactory, which is uses to resolve nodes.
 // TODO: now that this is more modular, try to unify this code with the
-//       the resolvers in namesys
-type Resolver struct {
+//       the resolvers in namesys.
+type BasicResolver struct {
```

Comment:
Reposting what was said on the discord for history.

I preffer the default implementation to be private because:
- Forcing everyone to use the interface else the code wont build and avoid over reliance of some code on the default implementation
- Loudly fails at compile time when doing an upgrade that you now need to use the interface.